File: clients/zig/src/cac.zig
//!
//! Wrapper implementation around the CAC (Context Aware Config) client library.
//!

// Imports
// ---------------------------------------------------------------------------

const std = @import("std");
const cac_bindings = @cImport({
    @cInclude("libcac_client.h");
});

const parse_json = @import("utils.zig").parse_json;
const parse_filter_keys = @import("utils.zig").parse_filter_keys;

const ClientError = @import("type.zig").ClientError;
pub const SetupConfig = @import("type.zig").SetupConfig;

pub const MergeStrategy = enum {
    MERGE,
    REPLACE,
};

// Types
// ---------------------------------------------------------------------------

pub const Client = struct {
    config: *const SetupConfig,
    ///
    /// [`client`] is a pointer to the CAC client instance.
    /// This is created when the cac client library is initialized. (see [`initCacClient`])
    ///
    client: *cac_bindings.Arc_Client,

    ///
    /// [`getClient`] returns a new client instance.
    ///
    /// Safety: This client as null pointer checks in place; if the client is not initialized the function returns an error.
    ///
    fn getClient(config: SetupConfig) ClientError!Client {
        const unsafe_client = cac_bindings.cac_get_client(config.tenant.ptr);

        if (unsafe_client) |safe_client| {
            return Client{ .client = safe_client, .config = &config };
        } else {
            return ClientError.ClientNotInitialized;
        }
    }

    ///
    /// [`deinit`] frees the client instance.
    ///
    /// consider using this function to defer the cleanup of the client instance.
    ///
    ///
    /// ```example
    ///
    /// const client = try Client.init(config);
    /// defer client.deinit();
    ///
    /// ```
    ///
    pub fn deinit(self: *const Client) void {
        cac_bindings.cac_free_client(self.client);
    }

    ///
    /// [`init`] initializes the CAC client.
    ///
    /// This client is initialized outside the zig scope and is accessed via the [`getClient`] function.
    /// This is done to manage a async runtime for managing cache and other artifacts optimizing the client.
    ///
    pub fn init(config: SetupConfig) ClientError!Client {
        const output = cac_bindings.cac_new_client(config.tenant.ptr, config.update_frequency, config.hostname.ptr);

        if (output == 1) {
            return ClientError.ClientInitializationError;
        }

        return try Client.getClient(config);
    }

    ///
    /// [`poll`] starts polling the CAC client for updates.
    ///
    /// This function is used to start polling the CAC client for updates.
    /// This is a blocking call, make sure to run this in a separate thread. The polling frequency is set in the [`StartupConfig`] struct using `update_frequency`.
    ///
    pub fn poll(self: *const Client) void {
        cac_bindings.cac_start_polling_update(self.config.tenant.ptr);
    }

    pub fn lastModifiedRaw(self: *const Client) ClientError![]const u8 {
        const c_string = cac_bindings.cac_get_last_modified(self.client);

        const string: [:0]const u8 = std.mem.span(c_string);

        return string;
    }

    fn getConfigRaw(self: *const Client, query: []const u8, prefix: []const u8) ClientError![]const u8 {
        const c_string = cac_bindings.cac_get_config(self.client, query.ptr, prefix.ptr);

        if (c_string == null) {
            return ClientError.FailedToGetConfig;
        }

        const string: [:0]const u8 = std.mem.span(c_string);

        return string;
    }

    fn getResolvedConfigRaw(self: *const Client, query: []const u8, filter_keys: []const u8, merge_strategy: []const u8) ClientError![]const u8 {
        const c_string = cac_bindings.cac_get_resolved_config(self.client, query.ptr, filter_keys.ptr, merge_strategy.ptr);

        if (c_string == null) {
            return ClientError.FailedToGetConfig;
        }

        const string: [:0]const u8 = std.mem.span(c_string);

        return string;
    }

    fn getDefaultConfigRaw(self: *const Client, filter_keys: []const u8) ClientError![]const u8 {
        const c_string = cac_bindings.cac_get_default_config(self.client, filter_keys.ptr);

        if (c_string == null) {
            return ClientError.FailedToGetConfig;
        }

        const string: [:0]const u8 = std.mem.span(c_string);

        return string;
    }

    // Allocation Based Functions

    pub fn getDefaultConfig(self: *const Client, comptime T: type, alloc: std.mem.Allocator, keys: []const []const u8) !T {
        const output = try self.getDefaultConfigRaw(try parse_filter_keys(alloc, keys, "|"));

        const config = try parse_json(T, alloc, output);

        freeFfiStringRaw(output);

        return config;
    }

    pub fn getResolvedConfig(self: *const Client, comptime O: type, alloc: std.mem.Allocator, query: anytype, filter_keys: []const []const u8, merge_strategy: MergeStrategy) !O {
        const query_ptr = try std.json.stringifyAlloc(alloc, query, .{});

        const output = try self.getResolvedConfigRaw(query_ptr, try parse_filter_keys(alloc, filter_keys, "|"), merge_strategy);

        const config = try parse_json(O, alloc, output);

        freeFfiStringRaw(output);

        return config;
    }

    pub fn getConfig(self: *const Client, comptime O: type, alloc: std.mem.Allocator, query: anytype, prefix: []const []const u8) !O {
        const prefix_ptr = try parse_filter_keys(alloc, prefix, ",");

        const query_ptr = try std.json.stringifyAlloc(alloc, query, .{});

        std.debug.print("Query: {s}\n", .{query_ptr});

        const output = try self.getConfigRaw(query_ptr, prefix_ptr);

        const config = try parse_json(O, alloc, output);

        freeFfiStringRaw(output);

        return config;
    }
};

///
///
/// [`freeFfiStringRaw`] frees the memory allocated by the CAC client.
///
/// As the memory allocations are done by the CAC client, it is important to free the memory allocated by the client.
///
/// Used to construct safe abstractions around the raw functions. (This can be used to copy over data to a memory space allocated by the provided allocator)
///
fn freeFfiStringRaw(_: []const u8) void {

    // BUG: The current implementation of `cac_free_string` accepts `char *` but a significant other functions defined return `const char *`. This makes it impossible to free the memory allocated by the CAC client.
    // cac_bindings.cac_free_string(s_ptr.ptr);
}

// Functions
// ---------------------------------------------------------------------------

test "test init" {
    const config = SetupConfig{ .tenant = "dev", .update_frequency = 10, .hostname = "http://localhost:8080" };

    const client = try Client.init(config);
    defer client.deinit();
}

test "test lastModifiedRaw" {
    const config = SetupConfig{ .tenant = "dev", .update_frequency = 10, .hostname = "http://localhost:8080" };

    const client = try Client.init(config);
    defer client.deinit();

    const last_modified = try client.lastModifiedRaw();

    try std.testing.expectEqualStrings(last_modified, "1970-01-01 00:00:00 UTC");
}

File: clients/swift/cac/cac.swift
import Foundation

typealias UnknownClientPointer = OpaquePointer
typealias Value = [String: Any]

enum MergeStrategy {
    case MERGE
    case REPLACE

    var show: String {
        switch self {
        case .MERGE:
            return "MERGE"
        case .REPLACE:
            return "REPLACE"
        }
    }
}

func createCacClient(tenant: String, frequency: UInt, hostname: String) -> Bool {
    return tenant.withCString { t -> Bool in
        return hostname.withCString { h -> Bool in
            let resp = cac_new_client(t, frequency, h);
            if (resp == 0) {
                return true
            }
            return false
        }
    }
}

func getCacClient(tenant: String) -> UnknownClientPointer? {
    return tenant.withCString { t -> UnknownClientPointer? in
        return cac_get_client(t)
    }
}

func cacStartPolling(tenant: String) {
    tenant.withCString { t in
        cac_start_polling_update(t)
    }
}

func getCacLastModified(client: UnknownClientPointer) -> String? {
    let resp = cac_get_last_modified(client)
    return resp.map { String(cString: $0) }
}

func parseJson(jsonString: String) -> Value? {
    if let jsonData = jsonString.data(using: .utf8) {
        do {
            return try JSONSerialization.jsonObject(with: jsonData, options: []) as? Value
        } catch {
            return nil
        }
    }
    return nil
}

func getResolvedConfig(client: UnknownClientPointer, context: String, filterKeys: [String]? = nil) -> Value? {
    let keys = filterKeys.map { $0.joined(separator: "|") }

    return context.withCString { c -> Value? in
        return MergeStrategy.MERGE.show.withCString { m -> Value? in
            let rawData : UnsafePointer<CChar>?
            if let k = keys {
                rawData = k.withCString { ck -> UnsafePointer<CChar>? in
                    return cac_get_resolved_config(client, c, ck, m)
                }
            } else {
                rawData = cac_get_resolved_config(client, c, nil, m)
            }
            return rawData.map { String(cString: $0) }.flatMap { parseJson(jsonString: $0) }
        }
    }
}

func getDefaultConfig(client: UnknownClientPointer, filterKeys: [String]? = nil) -> Value? {
    let keys = filterKeys.map { $0.joined(separator: "|") }
    let rawData : UnsafePointer<CChar>?
    if let k = keys {
        rawData = k.withCString { ck -> UnsafePointer<CChar>? in
            return cac_get_default_config(client, ck)
        }
    } else {
        rawData = cac_get_default_config(client, nil)
    }
    return rawData.map { String(cString: $0) }.flatMap { parseJson(jsonString: $0) }
}

func getConfig(client: UnknownClientPointer, filterQuery: [String]?, filterPrefix: [String]?) -> Value? {
    let query = filterQuery.map { $0.joined(separator: "|") }
    let prefix = filterPrefix.map { $0.joined(separator: "|") }
    let rawData : UnsafePointer<CChar>?

    if let q = query {
        rawData =
            q.withCString { qs -> UnsafePointer<CChar>? in
                if let p = prefix {
                    return p.withCString { ps -> UnsafePointer<CChar>? in
                        return cac_get_config(client, qs, ps)
                    }
                } else {
                    return cac_get_config(client, qs, nil)
                }
            }
    } else {
        if let p = prefix {
            rawData = p.withCString { ps -> UnsafePointer<CChar>? in
                return cac_get_config(client, nil, ps)
            }
        } else {
            rawData = cac_get_config(client, nil, nil)
        }
    }
    return rawData.map { String(cString: $0) }.flatMap { parseJson(jsonString: $0) }
}

func cacFreeClient(client: UnknownClientPointer) {
    cac_free_client(client)
}

func cacLastErrorMessage() -> String? {
    return cac_last_error_message().map { String(cString: $0) }
}
